# enforce dot notation whenever possible (`dot-notation`)

## Rule Details

This rule extends the base [`eslint/dot-notation`](https://eslint.org/docs/rules/dot-notation) rule.
It adds support for optionally ignoring computed `private` member access.

## How to use

<<<<<<< HEAD
```cjson
=======
```jsonc
>>>>>>> b459a022f47e1c55fb538e6c6b01f47908ccd92f
{
  // note you must disable the base rule as it can report incorrect errors
  "dot-notation": "off",
  "@typescript-eslint/dot-notation": ["error"]
}
```

## Options

See [`eslint/dot-notation`](https://eslint.org/docs/rules/dot-notation#options) options.
This rule adds the following options:

```ts
interface Options extends BaseDotNotationOptions {
  allowPrivateClassPropertyAccess?: boolean;
<<<<<<< HEAD
=======
  allowProtectedClassPropertyAccess?: boolean;
>>>>>>> b459a022f47e1c55fb538e6c6b01f47908ccd92f
}
const defaultOptions: Options = {
  ...baseDotNotationDefaultOptions,
  allowPrivateClassPropertyAccess: false,
<<<<<<< HEAD
=======
  allowProtectedClassPropertyAccess: false,
>>>>>>> b459a022f47e1c55fb538e6c6b01f47908ccd92f
};
```

### `allowPrivateClassPropertyAccess`

Example of a correct code when `allowPrivateClassPropertyAccess` is set to `true`

```ts
class X {
  private priv_prop = 123;
}

const x = new X();
x['priv_prop'] = 123;
```

<<<<<<< HEAD
=======
### `allowProtectedClassPropertyAccess`

Example of a correct code when `allowProtectedClassPropertyAccess` is set to `true`

```ts
class X {
  protected protected_prop = 123;
}

const x = new X();
x['protected_prop'] = 123;
```

>>>>>>> b459a022f47e1c55fb538e6c6b01f47908ccd92f
<sup>Taken with ❤️ [from ESLint core](https://github.com/eslint/eslint/blob/master/docs/rules/dot-notation.md)</sup>
